Toolstack support for ACPI shutdown without tools.

crogers1 opened the pull request:
domain.ml:
  * In hvm case, only do hard shutdown if guest is in s3. ACPI support
    now allows guest to poweroff without PV drivers.

vmact.ml:
  * Add extra case to do reboot if Xal reports Halt and toolstack reports
    reboot.  ACPI has no concept of 'reboot', so 'halt' status is returned
    after the domain is shutdown. In this unique case, don't report it as
    an error.

OXT-212

Signed-off-by: Chris Rogers <rogersc@ainfosec.com>

jean-edouard commented:
The if condition should be included in the pattern match.

crogers1 commented:
Good catch, that's the whole point of it being there...

jean-edouard commented:
I gave this a bunch of testing today.

While it works fine without tools installed, the behavior gets interesting with tools installed.

For example, if you configure a Windows guest with tools to hibernate-on-power-button, it will try to both hibernate and shutdown when the toolstack requests a shutdown (I'm guessing the former is caused by your patch while the latter is triggered by the tools). Then end result is that the guest hibernates, and then shuts down after waking up...

There's two solutions, either disable the codepath you added when tools are installed, or disable the power stuff in the guest tools. The latter is appealing, but limits the functionality to 2 guest-configured buttons, transferring some control from the platform to the guest... The former is probably way easier to implement too, so I think that's the way to go.

crogers1 commented:
This could also be a side effect of writing to the same xenstore node.  The watch I added in qemu waits on "control/shutdown" to be set and then it fires off a power button SCI to the guest.  If tools are installed, they also have a watch on "control/shutdown", and do whatever power event is defined.  Maybe if I have the no tools case watching a different xenstore nodes, we won't get this double power button behavior.

rossphilipson commented:
Long conversations about testing and whatnot should really be in the ticket IMO. E.g. stuff like this would be lost if you decided to close out the PR and do a new one.

Just throwing that out there...

stacktrust commented:
Discussion is best on tickets and we can also link/copy relevant PR discussion to the ticket.  It may be possible to do this automatically with the github API.

crogers1 commented:
Updated PR.  We don't share the same xenstore node that the PV drivers inspect to avoid multiple power operations on a single shutdown command.  Addressed Jed's issue with shutdown-after-hibernate problem.

crogers1 commented:
Updated PR again, retry shutdown command up to 5 times in case guest was not in a state to act on ACPI events.

cjp256 commented on xenvm/vmact.ml:
```diff
@@ -904,11 +904,27 @@ let shutdown_vm xc xs xal state force reason =
 
-		info "vm acknowledged shutdown request. waiting release";
-		(* wait domain release with xal *)
+			info "vm acknowledged shutdown request. waiting release";
```
Spacing?

cjp256 commented on xenvm/vmact.ml:
```diff
@@ -904,11 +904,27 @@ let shutdown_vm xc xs xal state force reason =
+						if force then (
+							Domain.hard_shutdown ~xc domid reason; true
+						) else (
+							Domain.shutdown_ack ~timeout:30. ~xc ~xs domid reason
```
Out of curiousity, what does this do?

crogers1 replied:
Xal.wait_release checks to see if the domain is dead for duration of timeout (30 sec). If we reached the timeout, try issuing the shutdown command again.  Ocaml doesn't have 'break' so if the domain is dead, wait_release returns immediately and no additional shutdowns are issued, but the loop is run to completion.

jean-edouard commented:
Do we know why during early boot guests don't support power button events? It works on bare-metal...
Also couldn't qemu tell us if a given acpi event has been "consumed"?

About the code, I'm not crazy about arbitrary timeouts, but if the answers to my previous questions are negative, then I guess we don't have a choice.

Looking at the code how it is right now, why do we do:

```
for ...
    let function = ...
    in function()
```

instead of

```
let function = ...
in for ...
    function()
```

?

rossphilipson commented:
Different stuffs happens on bare metal through firmware and circuitry prior to the OSPM being up and running, handling the power button. The answer to Jed's question is yes, we can know, in QEMU when the guest OS (the OSPM) enables the GPE bit effectively turning on event processing for the power button. We could write an XS node at this transition. The toolstack can make a decision to force kill the VM if the node is not set or do an ACPI shutdown if it is.

crogers1 commented:
Where would we implement that change to qemu?

rossphilipson commented:
If you want me to do it I would be more than happy. I know exactly where it needs to go.

crogers1 commented:
By all means, go ahead.  I'm not having much luck with looping/retrying shutdowns.  Either it's ugly, or it doesn't work right.  If I could key off that XS node, that would solve both problems.

rossphilipson commented:
Wilco. Where are you putting the shutdown node for the VM? I could just write a shutdown button enabled node next to it.

crogers1 commented:
Uh, the full path is /local/domain/<domid>/control/hvm-shutdown

rossphilipson commented:
Tested +1 merging.
